Sources/TuistLoader/Models+ManifestMappers/Headers+ManifestMapper.swift

```diff
@@ -12,58 +12,142 @@ extension TuistGraph.Headers {
-            privateHeaders = try resolveHeaders(manifest.private)
+            privateHeaders = try unfold(manifest.private)
             autoExlcudedPaths.formUnion(privateHeaders)
-            projectHeaders = try resolveHeaders(manifest.project)
+            projectHeaders = try unfold(manifest.project)
 
         case .publicExcludesPrivateAndProject:
-            projectHeaders = try resolveHeaders(manifest.project)
+            projectHeaders = try unfold(manifest.project)
             autoExlcudedPaths.formUnion(projectHeaders)
-            privateHeaders = try resolveHeaders(manifest.private)
+            privateHeaders = try unfold(manifest.private)
             autoExlcudedPaths.formUnion(privateHeaders)
-            publicHeaders = try resolveHeaders(manifest.public)
+            publicHeaders = try unfold(manifest.public, headersFromUmbrella: headersFromUmbrella)
         }
 
         return Headers(public: publicHeaders, private: privateHeaders, project: projectHeaders)
     }
 }
+
+extension FileList {
+    fileprivate func unfold(
+        generatorPaths: GeneratorPaths,
+        basenameFilter: Set<String>? = nil,
+        additionalExcluding: Set<AbsolutePath>
+    ) throws -> [AbsolutePath] {
+        return try globs.flatMap {
+            try $0.unfold(
+                generatorPaths: generatorPaths,
+                basenameFilter: basenameFilter,
+                additionalExcluding: additionalExcluding
+            )
+        }
+    }
+}
+
+extension FileListGlob {
+    fileprivate func unfold(
+        generatorPaths: GeneratorPaths,
+        basenameFilter: Set<String>?,
+        additionalExcluding: Set<AbsolutePath>
+    ) throws -> [AbsolutePath] {
+        let resolvedPath = try generatorPaths.resolve(path: glob)
+        let resolvedExcluding = (try resolvedExcluding(generatorPaths: generatorPaths)).union(additionalExcluding)
+        return resolvedPath.unfold(basenameFilter: basenameFilter, excluding: resolvedExcluding)
+    }
+
+    fileprivate func resolvedExcluding(generatorPaths: GeneratorPaths) throws -> Set<AbsolutePath> {
+        guard !excluding.isEmpty else { return [] }
+        var result: Set<AbsolutePath> = []
+        try excluding.forEach { path in
+            let resolved = try generatorPaths.resolve(path: path).pathString
+            let absolute = AbsolutePath(resolved)
+            let globs = AbsolutePath(absolute.dirname).glob(absolute.basename)
+            result.formUnion(globs)
+        }
+        return result
+    }
+}
+
+extension AbsolutePath {
+    fileprivate func unfold(basenameFilter: Set<String>?, excluding: Set<AbsolutePath>) -> [AbsolutePath] {
+        FileHandler.shared.glob(AbsolutePath.root, glob: String(pathString.dropFirst())).filter {
+            guard let fileExtension = $0.extension else {
+                return false
+            }
+            guard TuistGraph.Headers.extensions.contains(".\(fileExtension)"),
```

Comment:
I'd extract this method into [this extension](https://github.com/tuist/tuist/blob/main/Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift) and make it generic (decouple it from `TuistGraph.Headers.extensions`) so that we can reuse it from other places.

Reply:
Moreover I'd add unit test for all the different code paths we have in the method.

Reply:
ugh, ok. it's not something new - it was inside main function, I just extracted to extensions, to have more clear logic. but ok - will move and add unit tests